app/controllers/users_controller.rb

```diff
@@ -496,15 +496,29 @@ def is_local_username
     usernames -= groups
     usernames.each(&:downcase!)
 
-    cannot_see = []
+    cannot_see = {}
     here_count = nil
 
     topic_id = params[:topic_id]
     if topic_id.present? && topic = Topic.find_by(id: topic_id)
       usernames.each do |username|
-        if !Guardian.new(User.find_by_username(username)).can_see?(topic)
-          cannot_see.push(username)
+        user = User.find_by_username(username)
+        next if user.blank?
+
+        cannot_see_reason = nil
+        if !Guardian.new(user).can_see?(topic)
+          cannot_see_reason = topic.private_message? ? :private : :category
+        elsif TopicUser.exists?(topic: topic, user: user, notification_level: TopicUser.notification_levels[:muted])
```

Comment:
I have the same concern here as https://github.com/discourse/discourse/pull/15339/files#r771072939. We're now hitting the DB multiple times per user/username to check. Will we be able to preload these while loading the user records?